File: shared/Ai/AgeBand.php
<?php

declare(strict_types=1);

namespace Kidepik\Shared\Ai;

/**
 * Bandas pedagógicas abiertas (SPEC_APP_AGE_BANDS).
 */
final class AgeBand
{
    public const EARLY = 'band_early';
    public const CHILD = 'band_child';
    public const TWEEN = 'band_tween';
    public const TEEN = 'band_teen';
    public const ADULT = 'band_adult';
    public const SENIOR = 'band_senior';

    /** @var list<string> */
    public const ALL = [
        self::EARLY,
        self::CHILD,
        self::TWEEN,
        self::TEEN,
        self::ADULT,
        self::SENIOR,
    ];

    public static function fromAgeYears(int $age): string
    {
        return match (true) {
            $age <= 7 => self::EARLY,
            $age <= 10 => self::CHILD,
            $age <= 13 => self::TWEEN,
            $age <= 17 => self::TEEN,
            $age <= 64 => self::ADULT,
            default => self::SENIOR,
        };
    }

    public static function isValid(?string $band): bool
    {
        return $band !== null && in_array($band, self::ALL, true);
    }

    /** Migra ids legacy age_7 / age_9. */
    public static function fromLegacy(?string $legacy, ?int $ageYears = null): ?string
    {
        if ($legacy === null || $legacy === '') {
            return $ageYears !== null ? self::fromAgeYears($ageYears) : null;
        }
        if (in_array($legacy, self::ALL, true)) {
            return $legacy;
        }
        if ($legacy === 'age_7') {
            return ($ageYears !== null && $ageYears <= 7) ? self::EARLY : self::CHILD;
        }
        if ($legacy === 'age_9') {
            return ($ageYears !== null && $ageYears >= 11) ? self::TWEEN : self::CHILD;
        }

        return null;
    }

    public static function assertAgeYears(int $age): void
    {
        if ($age < 5 || $age > 99) {
            throw new \InvalidArgumentException('age_years invalid');
        }
    }
}

File: backend/app/catalogs/age_band.py
from __future__ import annotations


class AgeBand:
    EARLY = "band_early"
    CHILD = "band_child"
    TWEEN = "band_tween"
    TEEN = "band_teen"
    ADULT = "band_adult"
    SENIOR = "band_senior"
    ALL = (EARLY, CHILD, TWEEN, TEEN, ADULT, SENIOR)

    @classmethod
    def from_age_years(cls, age: int) -> str:
        if age <= 7:
            return cls.EARLY
        if age <= 10:
            return cls.CHILD
        if age <= 13:
            return cls.TWEEN
        if age <= 17:
            return cls.TEEN
        if age <= 64:
            return cls.ADULT
        return cls.SENIOR

    @classmethod
    def is_valid(cls, band: str | None) -> bool:
        return band in cls.ALL

    @classmethod
    def from_legacy(cls, legacy: str | None, age_years: int | None = None) -> str | None:
        if not legacy:
            return cls.from_age_years(age_years) if age_years is not None else None
        if legacy in cls.ALL:
            return legacy
        if legacy == "age_7":
            return cls.EARLY if age_years is not None and age_years <= 7 else cls.CHILD
        if legacy == "age_9":
            return cls.TWEEN if age_years is not None and age_years >= 11 else cls.CHILD
        return None

    @staticmethod
    def assert_age_years(age: int) -> None:
        if age < 5 or age > 99:
            raise ValueError("age_years invalid")
